### PR TITLE
Adding OAEP Data Migration client to DAS 

### PR DESCRIPTION
**Pull request by AnuGayan**
## Purpose
In DAS 3.2.0 crypto util uses a modified crypto algorithm. Due to this reason when migrating from DAS 3.1.0, several files and passwords should be migrated to the new algorithm in order to work with new DAS 3.2.0.

### CODE REVIEW

File: modules/migration/oaep-data-migration/pom.xml [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <groupId>org.wso2.das</groupId>
+        <artifactId>wso2das-parent</artifactId>
+        <version>3.2.1-SNAPSHOT</version>
+        <relativePath>../../../pom.xml</relativePath>
+    </parent>
+
+    <groupId>org.wso2.carbon.das</groupId>
```
Comment (tishan89) [inline modules/migration/oaep-data-migration/pom.xml]: fix group id

File: modules/migration/oaep-data-migration/pom.xml [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    <artifactId>org.wso2.das.migration</artifactId>
+    <packaging>bundle</packaging>
+    <name>org.wso2.carbon.das.migration</name>
+    <modelVersion>4.0.0</modelVersion>
+
+    <version>1.0-SNAPSHOT</version>
```
Comment (tishan89) [inline modules/migration/oaep-data-migration/pom.xml]: fix version

File: modules/migration/oaep-data-migration/pom.xml [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            <artifactId>org.wso2.carbon.logging</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>commons-logging</groupId>
+            <artifactId>commons-logging</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon</groupId>
+            <artifactId>org.wso2.carbon.core.services</artifactId>
+            <version>${carbon.kernel.version}</version>
```
Comment (tishan89) [inline modules/migration/oaep-data-migration/pom.xml]: move versions to parent

File: modules/migration/oaep-data-migration/pom.xml [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Export-Package>
+                            org.wso2.migration.*,
+                            org.wso2.carbon.core.*
+                        </Export-Package>
+                        <Private-Package>
+                            org.wso2.carbon.analytics.common.jmx.agent.profiles.*
+                        </Private-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (tishan89) [inline modules/migration/oaep-data-migration/pom.xml]: remove dynamic import

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/DataMigrationClientImpl.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    public void execute() throws DataMigrationException {
+        try {
+            MigrationHolder migrationHolder = MigrationHolder.getInstance();
+            List<Migrator> migrators = migrationHolder.getMigrators();
+            for (Migrator migrator : migrators) {
+                migrator.migrate();
+            }
+            LOG.info("Migration was successful.");
+        } catch (Exception e) {
+            throw new DataMigrationException("Error occurred while migrating. Migration stopped. ", e);
```
Comment (tishan89) [inline modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/DataMigrationClientImpl.java]: Can we be more specific with exceptions

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/internal/MigrationServiceComponent.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+     */
+    protected void activate(ComponentContext context) {
+        try {
+            BundleContext bundleContext = context.getBundleContext();
+            bundleContext.registerService(DataMigrationClient.class, new DataMigrationClientImpl(), null);
+
+            // if -Dmigrate option is used.
+            String migrate = System.getProperty("migrate");
+            if (Boolean.parseBoolean(migrate)) {
+                LOG.info("Executing Migration client : " + DataMigrationClient.class.getName());
```
Comment (tishan89) [inline modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/internal/MigrationServiceComponent.java]: executing or initializing?

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/service/InputOutputDataMigration.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+    @Override
+    public void migrate() throws DataMigrationException {
+        String carbonPath = System.getProperty(DataMigrationConstants.CARBON_HOME);
+        try {
+            migratePublishers(carbonPath);
+            migrateReceivers(carbonPath);
+
+        } catch (Exception e) {
+            throw new DataMigrationException("Error while migrating data : ", e);
```
Comment (tishan89) [inline modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/service/InputOutputDataMigration.java]: can we be more specific with exception handling?

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/service/KeyStoreAndTrustStoreMigration.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        carbonContext.setTenantDomain(tenant.getDomain());
+    }
+
+    private void migrateKeystorePassword() throws DataMigrationException, RegistryException, CryptoException,
+            UserStoreException {
+        try {
+            migrateKeyStorePasswordForTenant(SUPER_TENANT_ID);
+            LOG.info("Keystore passwords migrated for tenant : " + SUPER_TENANT_DOMAIN_NAME);
+        } catch (Exception e) {
+            LOG.error("Error while migrating Keystore passwords for tenant : " + SUPER_TENANT_DOMAIN_NAME);
```
Comment (tishan89) [inline modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/service/KeyStoreAndTrustStoreMigration.java]: why log and throw?

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/service/KeyStoreAndTrustStoreMigration.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+        //migrating tenant configurations
+        Tenant[] tenants = MigrationServiceDataHolder.getRealmService().getTenantManager().getAllTenants();
+        for (Tenant tenant : tenants) {
+            try {
+                startTenantFlow(tenant);
+                migrateKeyStorePasswordForTenant(tenant.getId());
+                LOG.info("Keystore passwords migrated for tenant : " + tenant.getDomain());
+            } catch (Exception e) {
+                LOG.error("Error while migrating keystore passwords for tenant : " + tenant.getDomain());
```
Comment (tishan89) [inline modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/service/KeyStoreAndTrustStoreMigration.java]: why log and throw?

File: modules/migration/oaep-data-migration/pom.xml (hunk 1 above)
Comment (AnuGayan) [reply modules/migration/oaep-data-migration/pom.xml]: fixed in #315

File: modules/migration/oaep-data-migration/pom.xml (hunk 2 above)
Comment (AnuGayan) [reply modules/migration/oaep-data-migration/pom.xml]: fixed by #315

File: modules/migration/oaep-data-migration/pom.xml (hunk 3 above)
Comment (AnuGayan) [reply modules/migration/oaep-data-migration/pom.xml]: fixed by #315

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/DataMigrationClientImpl.java (hunk 5 above)
Comment (AnuGayan) [reply modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/DataMigrationClientImpl.java]: fixed by #315

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/internal/MigrationServiceComponent.java (hunk 6 above)
Comment (AnuGayan) [reply modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/internal/MigrationServiceComponent.java]: in here we are executing the client after initializing it..

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/service/InputOutputDataMigration.java (hunk 7 above)
Comment (AnuGayan) [reply modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/service/InputOutputDataMigration.java]: fixed by #315

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/service/KeyStoreAndTrustStoreMigration.java (hunk 9 above)
Comment (AnuGayan) [reply modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/service/KeyStoreAndTrustStoreMigration.java]: fixed by #331

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/service/KeyStoreAndTrustStoreMigration.java (hunk 8 above)
Comment (AnuGayan) [reply modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/service/KeyStoreAndTrustStoreMigration.java]: fixed by #331

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/DataMigrationClientImpl.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+import org.wso2.migration.exception.DataMigrationException;
+import org.wso2.migration.service.Migrator;
+
+import java.util.List;
+
+/**
+ * Implementation of Data migration client.
+ **/
+public class DataMigrationClientImpl implements DataMigrationClient {
+    private static final Log LOG = LogFactory.getLog(DataMigrationClientImpl.class);
```
Comment (pcnfernando) [inline modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/DataMigrationClientImpl.java]: The logger reference is not a constant, but a final reference, and should not be in uppercase. Check all places

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/MigrationHolder.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Migration data holder class.
+ **/
+public class MigrationHolder {
+    private static MigrationHolder migrationHolder = new MigrationHolder();
+    private List<Migrator> migrators = new ArrayList<>();
+
```
Comment (pcnfernando) [inline modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/MigrationHolder.java]: Remove unnecessary new lines. Check all places

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/DataMigrationClientImpl.java (hunk 10 above)
Comment (AnuGayan) [reply modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/DataMigrationClientImpl.java]: Fixed by 325baaba7cbaefc9ed303543ab90a17e859fcc1a

File: modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/MigrationHolder.java (hunk 11 above)
Comment (AnuGayan) [reply modules/migration/oaep-data-migration/src/main/java/org/wso2/migration/MigrationHolder.java]: Fixed by 325baaba7cbaefc9ed303543ab90a17e859fcc1a

File: modules/migration/oaep-data-migration/pom.xml (hunk 4 above)
Comment (AnuGayan) [reply modules/migration/oaep-data-migration/pom.xml]: Removed by 8839c06f8bd6789cbe9a2d203cf2232e05c645f3

**Pull request merged**